duo.c

```
@@ -138,18 +138,39 @@ _sign_request(struct duo_ctx *ctx, const char *method, const char *uri,
     const char *qs)
 {
         BIO *bio, *b64;
-	HMAC_CTX hmac;
-	unsigned char MD[SHA_DIGEST_LENGTH];
+#if OPENSSL_VERSION_NUMBER >= 0x1010000fL
```

Thanks for getting a PR started on this. I am wondering what you think about taking an approach a little more similar to what is done in https://github.com/duosecurity/duo_unix/commit/673b3ce97763530d09f8a2002959b6b45214ac1a

Essentially the major difference between your PR and that PR is that we implemented the `new` and `free` functions if your version of openssl is old. That way we could use the new interface regardless of version. This means the large `if` `else` macros below can be consolidated because we don't need to worry about whether or not `HMAC_CTX` is the full object or just a pointer. It will always be a pointer.

---

You're welcome. The approach in `duo_unix` is certainly more elegant. I'll have a go at it. I don't see where `HMAC_Init()` deprecation in being handled in [duosecurity/duo_unix@673b3ce](https://github.com/duosecurity/duo_unix/commit/673b3ce97763530d09f8a2002959b6b45214ac1a). Shouldn't that have a similar function in the `if/else` macro block for versions < 1.1?

---

I actually didn't realize HMAC_Init was deprecated. Unless I'm mistaken deprecated doesn't mean it is removed so we can still get by using it for now until it's actually removed in a future release. However getting ahead of the curve here is not a bad idea.

---

If I'm reading the [OpenSSL wiki page](https://wiki.openssl.org/index.php/OpenSSL_1.1.0_Changes) correctly (under _Major Changes So Far_..), it did't get flagged for deprecation until 1.1.0 so it's kind of new. There's something in the wiki about `HMAC_Init()` (along with `HMAC_Cleanup()`) being "off by default" in 1.1.0.. not sure what that means exactly. I've added this to my `OPENSSL_VERSION_NUMBER` macro in `duo.c`.  The fifth parameter is silently ignored if complied on < 1.1.0 which is `NULL` in duo.c anyhow. Seem Ok?
```
...
static int
HMAC_Init_ex(HMAC_CTX *ctx, const void *key, int len, const EVP_MD *md, ENGINE *impl)
{
   /* older ssl version so ignore impl parameter which wouldn't be used anyway */
   return (HMAC_Init (ctx, key, len, md));
}
#endif
```

---

Whoops - nevermind. `HMAC_Init_ex()` is a pre 1.1.0 thing as well.